[Yuriy Yarovikov] 02-objects-arrays-intro-to-testing

YuraNick opened the pull request.

dosandk commented on 02-objects-arrays-intro-to-testing/1-create-getter/index.js:
```diff
@@ -4,5 +4,16 @@
+    
+    return (obj) => {
+        let result = obj;
+        pathArray.map( (item) => result = getValue(result, item) );
```
использование "map" здесь не обосновано. Да, "map" может итерировать массив, но его ключевое назначение - массив преобразовать и вернуть новый преобразованный. В данном случае лучше использовать либо "forEach" или "for ... of" для итерации

dosandk commented on 02-objects-arrays-intro-to-testing/2-invert-object/index.js:
```diff
@@ -4,5 +4,6 @@
 export function invertObj(obj) {
-
+    if (typeof obj !== 'object' || obj === null) return;
+    return Object.fromEntries( Object.entries(obj).map( ([key, value]) => [value, key] ) );
```
также, лучше заменить "Object.fromEntries" и "map" на один "reduce" - подробнее расскажу на занятии

dosandk commented on 02-objects-arrays-intro-to-testing/3-trim-symbols/index.js:
```diff
@@ -5,5 +5,20 @@
  * @returns {string} - the new string without extra symbols according passed size
  */
 export function trimSymbols(string, size) {
+    if (isNaN(size)) return string;
```
isNaN - здесь не нужен, нам нужно проверить "size" на соответствие нулю

YuraNick replied:
Без `isNaN(size)`, например `if (size === 0) `- не работает

dosandk commented on 02-objects-arrays-intro-to-testing/3-trim-symbols/index.js:
```diff
@@ -5,5 +5,20 @@
+    
+    return string.split('').filter( (item) => filterBySize(item, obj) ).join('');
+
+    function filterBySize(curLetter, setObj) {
```
решение прокоментирую на занятии

dosandk commented on 02-objects-arrays-intro-to-testing/4-uniq/index.js:
```diff
@@ -4,5 +4,6 @@
 export function uniq(arr) {
-
+    if (! Array.isArray(arr)) return [];
+    return [... new Set(arr)]; // это работает, но стоит ли так писать?
```
стоит, пробела только не должно быть между "..." и "new"

dosandk commented on 02-objects-arrays-intro-to-testing/4-uniq/index.js:
```diff
@@ -4,5 +4,6 @@
  */
 export function uniq(arr) {
-
+    if (! Array.isArray(arr)) return [];
```
эта проверка лишняя

YuraNick commented:
Спасибо!
 
Вот на примере этого момента (вставил чуть ниже жирным текстом) вопрос общего плана:
если создается например функция, которая по её описанию должна в качестве параметра получать определённого типа переменную, в данном примере это массив — то как по-вашему, нужно ли внутри функции проверять тип полученного функцией параметра, arr в данном примере? Ведь если будет передан, например, объект вместо массива — то будет ошибка.
Или это и правильно — пусть будет ошибка, если передали не тот параметр?
 
In   02-objects-arrays-intro-to-testing/4-uniq/index.js :
> @@ -4,5 +4,6 @@

  * @returns {*[]} - the new array with uniq values

  */

 export function uniq(arr) {

-

+    if (! Array.isArray(arr)) return [];

эта проверка лишняя
 
 
--------------------------
С уважением,
Яровиков Юрий
8-913-669-88-08
 
  
>Четверг, 21 мая 2020, 15:15 +07:00 от Vladimir Shevchuk <notifications@github.com>:
> 
> 
>@dosandk commented on this pull request.
>----------------------------------------------------------------------
>In  02-objects-arrays-intro-to-testing/1-create-getter/index.js :
>> @@ -4,5 +4,16 @@
>
>  * @returns {function} - function-getter which allow get value from object by set path
>
>  */
>
> export function createGetter(path) {
>
>+    const pathArray = path.split('.');
>
>+    
>
>+    return (obj) => {
>
>+        let result = obj;
>
>+        pathArray.map( (item) => result = getValue(result, item) );
>
>
>использование "map" здесь не обосновано. Да, "map" может итерировать массив, но его ключевое назначение - массив преобразовать и вернуть новый преобразованный. В данном случае лучше использовать либо "forEach" или "for ... of" для итерации
>----------------------------------------------------------------------
>In  02-objects-arrays-intro-to-testing/2-invert-object/index.js :
>> @@ -4,5 +4,6 @@
>
>  * @returns {object | undefined} - returns new object or undefined if nothing did't pass
>
>  */
>
> export function invertObj(obj) {
>
>-
>
>+    if (typeof obj !== 'object' || obj === null) return;
>
>+    return Object.fromEntries( Object.entries(obj).map( ([key, value]) => [value, key] ) );
>
>
>также, лучше заменить "Object.fromEntries" и "map" на один "reduce" - подробнее расскажу на занятии
>----------------------------------------------------------------------
>In  02-objects-arrays-intro-to-testing/3-trim-symbols/index.js :
>> @@ -5,5 +5,20 @@
>
>  * @returns {string} - the new string without extra symbols according passed size
>
>  */
>
> export function trimSymbols(string, size) {
>
>+    if (isNaN(size)) return string;
>
>
>isNaN - здесь не нужен, нам нужно проверить "size" на соответствие нулю
>----------------------------------------------------------------------
>In  02-objects-arrays-intro-to-testing/3-trim-symbols/index.js :
>> @@ -5,5 +5,20 @@
>
>  * @returns {string} - the new string without extra symbols according passed size
>
>  */
>
> export function trimSymbols(string, size) {
>
>+    if (isNaN(size)) return string;
>
>+
>
>+    const obj = {
>
>+        identicalCnt : 1,
>
>+        prevLetter : '',
>
>+        setMaxCnt : size
>
>+    };
>
>+    
>
>+    return string.split('').filter( (item) => filterBySize(item, obj) ).join('');
>
>+
>
>+    function filterBySize(curLetter, setObj) {
>
>
>решение прокоментирую на занятии
>----------------------------------------------------------------------
>In  02-objects-arrays-intro-to-testing/4-uniq/index.js :
>> @@ -4,5 +4,6 @@
>
>  * @returns {*[]} - the new array with uniq values
>
>  */
>
> export function uniq(arr) {
>
>-
>
>+    if (! Array.isArray(arr)) return [];
>
>+    return [... new Set(arr)]; // это работает, но стоит ли так писать?
>
>
>стоит, пробела только не должно быть между "..." и "new"
>----------------------------------------------------------------------
>In  02-objects-arrays-intro-to-testing/4-uniq/index.js :
>> @@ -4,5 +4,6 @@
>
>  * @returns {*[]} - the new array with uniq values
>
>  */
>
> export function uniq(arr) {
>
>-
>
>+    if (! Array.isArray(arr)) return [];
>
>
>эта проверка лишняя
>—
>You are receiving this because you authored the thread.
>Reply to this email directly,  view it on GitHub , or  unsubscribe .
 

YuraNick commented:
5-sum solved

dosandk commented on 02-objects-arrays-intro-to-testing/5-sum/index.js:
```diff
@@ -5,4 +5,25 @@
+
+    /** 
+     * хотя сам счетчик вот так можно было бы сделать:
+     * if (sum.sumVal === undefined) sum.sumVal = 0;
```
если хранить его как свойство ф-ции с этим будет много проблем

dosandk commented on 02-objects-arrays-intro-to-testing/5-sum/index.js:
```diff
@@ -5,4 +5,25 @@
+    */
+
+    calcSumm.sumVal = Number(n) || 0;
+    calcSumm[Symbol.toPrimitive] = () => calcSumm.sumVal;
```
good 👍
